From b2eb241c6c3dde5ff62f2fab1670dec232bc576d Mon Sep 17 00:00:00 2001
From: Sergey Temerkhanov <s.temerkhanov@gmail.com>
Date: Tue, 3 Nov 2015 19:20:50 -0800
Subject: [PATCH 0331/1239] cavium: Modify clock-frequency property based on
 external clock change

The UAA is initialized by the Boot Stub firmware.
The SCLK frequency can be changed and that affects the IBRD and FBRD
register values of the UAA. This eventually has a need to modify the
"clock-frequency" property so that the Linux PL011 is happy.

Signed-off-by: Radha Mohan Chintakuntla <rchintakuntla@cavium.com>
---
 board/cavium/thunderx/fdt.c | 28 ++++++++++++++++++++++++++++
 1 file changed, 28 insertions(+)

diff --git a/board/cavium/thunderx/fdt.c b/board/cavium/thunderx/fdt.c
index 571d652772..90b2e2a985 100644
--- a/board/cavium/thunderx/fdt.c
+++ b/board/cavium/thunderx/fdt.c
@@ -205,6 +205,33 @@ static void ft_setup_coremask(void *fdt)
 }
 #if defined(CONFIG_OF_BOARD_SETUP)
 
+static void ft_setup_uaa_clk(char *fdt)
+{
+	unsigned int ibrd, fbrd;
+	unsigned long baud = CONFIG_BAUDRATE;
+	uint32_t refclk;
+	int nodeoffset, err;
+
+
+	ibrd = CSR_READ_PA(0, UAAX_IBRD(0));
+	ibrd = ibrd & 0xffff;
+
+	fbrd = CSR_READ_PA(0, UAAX_FBRD(0));
+	fbrd = fbrd & 0x3f;
+
+	refclk = (baud * (64 * ibrd + fbrd)) / 4;
+
+	nodeoffset = fdt_path_offset(fdt, "/soc/refclkuaa");
+
+	err = fdt_setprop_u32(fdt, nodeoffset, "clock-frequency", refclk);
+
+	if(err < 0) {
+		printf("WARNING: could not set %s: %s.\n", "clock-frequency",
+				fdt_strerror(err));
+		return;
+	}
+}
+
 int ft_board_setup(void *blob, bd_t *bd)
 {
 	unsigned int node;
@@ -218,6 +245,7 @@ int ft_board_setup(void *blob, bd_t *bd)
 		ft_setup_macs(blob, node);
 	}
 #endif
+	ft_setup_uaa_clk(blob);
 
 	return 0;
 }
-- 
2.29.0

